File: Libraries/Nori2/resources/java-app/NoriHelper.java
package org.crayonlang.libraries.nori2;

import java.awt.Color;
import java.awt.Component;
import java.awt.Dimension;
import java.util.ArrayList;
import javax.swing.JButton;
import javax.swing.JComponent;
import javax.swing.JPanel;
import org.crayonlang.interpreter.structs.Value;
import org.crayonlang.interpreter.FastList;

final class NoriHelper {

	private NoriHelper() {}

	public static void addChildToParent(Object childObj, Object parentObj) {
		JPanel jp = (JPanel) parentObj;
		JComponent jc = (JComponent)  childObj;
		jp.add(jc, null);
	}

	public static void closeWindow(Object window) {
		throw new RuntimeException();
	}

	public static void ensureParentLinkOrder(Object parent, Object[] children) {
		Component[] actualChildren = ((JPanel)parent).getComponents();
		if (actualChildren.length == children.length) {
			int validUntil = 0;
			for (int i = 0; i < children.length; ++i) {
				if (actualChildren[i] == children[i]) {
					validUntil = i + 1;
				} else {
					break;
				}
			}
			if (validUntil == children.length) return;
		}
		throw new RuntimeException();
	}

	public static Object instantiateElement(int type, ElementProperties properties) {
		JComponent jc = null;
		switch (type) {
			case 1:
				jc = new JPanel();
				jc.setBackground(new Color(properties.bg_red, properties.bg_green, properties.bg_blue, properties.bg_alpha));
				break;
			case 3:
				JPanel jp = new JPanel();
				jp.setLayout(null);
				jc = jp;
				break;
			case 4:
				JButton btn = new JButton(properties.misc_string_0);
				jc = btn;
				break;
			default:
				throw new RuntimeException("not implemented");
		}
		
		jc.setBounds(properties.render_left, properties.render_top, properties.render_width, properties.render_height);
		
		return jc;
	}

	public static Object instantiateWindow(WindowProperties properties) {
		String title = properties.title;
		int width = properties.width;
		int height = properties.height;
		return new NoriWindow(title, width, height);
	}

	public static void invalidateElementProperty(int type, Object element, int key, Object value) {
		throw new RuntimeException();
	}

	public static void invalidateWindowProperty(Object window, int key, Object value) {
		throw new RuntimeException();
	}

	public static void showWindow(Object wObj, Object[] ignored, Object rootElement) {
		NoriWindow window = (NoriWindow) wObj;
		window.setContent((JComponent) rootElement);
		window.show();
	}

	public static void updateLayout(Object obj, int typeId, int x, int y, int width, int height) {
		JComponent jc = (JComponent) obj;
		jc.setBounds(x, y, width, height);
	}
}

File: Libraries/Nori2/resources/csharp-app/NoriHelper.cs
using Interpreter.Structs;
using System;
using System.Collections.Generic;

namespace Interpreter.Libraries.Nori2
{
    public static class NoriHelper
    {
        public static void AddChildToParent(object child, object parent)
        {
            ((System.Windows.Forms.Panel)parent).Controls.Add((System.Windows.Forms.Control)child);
        }

        public static void CloseWindow(object window)
        {
            throw new NotImplementedException();
        }

        public static void EnsureParentLinkOrder(object parent, object[] children)
        {
            System.Windows.Forms.Control.ControlCollection actualCollection = ((System.Windows.Forms.Control)parent).Controls;
            System.Windows.Forms.Control[] actualOrder = GetControls(actualCollection);

            int perfectMatchUntil = 0;
            if (actualOrder.Length == children.Length)
            {
                for (int i = 0; i < children.Length; ++i)
                {
                    if (children[i] == actualOrder[i])
                    {
                        perfectMatchUntil = i + 1;
                    }
                }
                if (perfectMatchUntil == children.Length)
                {
                    return;
                }
            }
            throw new NotImplementedException();
        }

        public static object InstantiateElement(int type, ElementProperties properties)
        {
            int left = properties.render_left;
            int top = properties.render_top;
            int width = properties.render_width;
            int height = properties.render_height;

            switch (type)
            {
                case 1: // Rectangle
                    Rectangle rect = new Rectangle();
                    rect.SetColor(
                        properties.bg_red,
                        properties.bg_green,
                        properties.bg_blue,
                        properties.bg_alpha);
                    return rect;

                case 2: // Canvas
                    Canvas canvas = new Canvas();
                    canvas.SetPosition(left, top, width, height);
                    return canvas;

                case 3: // ScrollPanel
                    ScrollPanel scrollPanel = new ScrollPanel();
                    scrollPanel.SetPosition(left, top, width, height);
                    return scrollPanel;

                case 4: // Button
                    Button button = new Button();
                    button.SetPosition(left, top, width, height);
                    button.Text = properties.misc_string_0;
                    return button;

                default:
                    throw new NotImplementedException();
            }
        }

        public static object InstantiateWindow(WindowProperties properties)
        {
            System.Windows.Forms.Form form = new System.Windows.Forms.Form();
            form.Text = properties.title;
            form.Size = new System.Drawing.Size(properties.width, properties.height);
            return form;
        }

        public static void InvalidateElementProperty(int type, object element, int key, object value)
        {
            throw new NotImplementedException();
        }

        public static void InvalidateWindowProperty(object window, int key, object value)
        {
            throw new NotImplementedException();
        }

        public static void ShowWindow(object window, object[] ignored, object rootElement)
        {
            System.Windows.Forms.Form form = (System.Windows.Forms.Form)window;
            form.Controls.Add((System.Windows.Forms.Panel)rootElement);
            form.ShowDialog();
        }

        public static System.Windows.Forms.Control[] GetControls(System.Windows.Forms.Control.ControlCollection cc)
        {
            List<System.Windows.Forms.Control> output = new List<System.Windows.Forms.Control>();
            foreach (System.Windows.Forms.Control c in cc)
            {
                output.Add(c);
            }
            return output.ToArray();
        }

        public static void UpdateLayout(object element, int typeId, int x, int y, int width, int height)
        {
            switch (typeId)
            {
                case 1:
                    Rectangle r = (Rectangle)element;
                    r.SetPosition(x, y, width, height);
                    break;

                case 3:
                    ScrollPanel sp = (ScrollPanel)element;
                    sp.SetPosition(x, y, width, height);
                    break;

                case 4:
                    Button btn = (Button)element;
                    btn.SetPosition(x, y, width, height);
                    break;

                default:
                    throw new NotImplementedException();
            }
        }
    }
}
